{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bdad84ef",
   "metadata": {},
   "source": [
    "# Estimate single-cell density at observed time point. (Optional)\n",
    "\n",
    "***Pseudodyanmics+*** leverage a concpet called Physics-Informed Neural Networks (PINN) which can solve the forward and inverse problem of PDE. We applies the continuous PINN which requires:\n",
    "- the model correctly predict density at observed time points and data points\n",
    "- the model follows the PDE (physics formular) anywhere (collocation point : unobserved cell state) and anytime (collocation time : unobserved time point)\n",
    "\n",
    "\n",
    "In this notebook, we will go through several density estimator that could served as the label for solving PINN. <font color='Blue'>\n",
    "1. Gaussian KDE\n",
    "2. Mellon (the latest)\n",
    "3. HBE (scalabl)\n",
    "\n",
    "</font>\n",
    "\n",
    "Remember that density estimation in high-dimensional space is not retrival. None of the method above captured the true density as the true manifold of the single-cell data is never known."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "22ba021f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "workding directory changed to: /Users/weizhongzheng/Documents/python_project/pseudodynamics_plus\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import os, sys\n",
    "if sys.platform.startswith(\"darwin\"):\n",
    "    os.environ['KMP_DUPLICATE_LIB_OK']='True'\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import scanpy as sc\n",
    "sc.settings.set_figure_params(frameon=False, dpi=30)\n",
    "\n",
    "import pseudodynamics as pdp\n",
    "os.chdir(pdp.main_dir)\n",
    "print(\"workding directory changed to:\", pdp.main_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "187b7dae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 49390 × 4814\n",
       "    obs: 'n_genes', 'n_counts', 'mt_count', 'mt_frac', 'doublet_scores', 'predicted_doublets', 'xist_logn', 'Ygene_logn', 'xist_bin', 'Ygene_bin', 'sex_adata', 'biosample_id', 'cellid', 'RBG', 'SLXid', 'index', '10xsample_description', 'sex_mixed', 'sex_meta', 'mouse_id', 'sortedcells', 'expected_cells_10x', 'cellranger_cellsfound', 'chemistry', 'tom', 'expdate', 'batch', 'timepoint_tx_days', 'start_age', 'sample_id', 'countfile', 'S_score', 'G2M_score', 'phase', 'leiden', 'SLX', 'plate_sorted', 'plate_rearranged', 'well_sorted', 'well_rearranged', 'set_index', 'CI_index', 'mouse_platelabel', 'sort_method', 'sample.name', 'population', 'sex', 'countfolder', 'batch_plate_sorted', 'data_type', 'sex_combined', 'longname', 'anno_man', 'leiden_DM', 'HSCscore', 'nn_HSCscore', 'isroot', 'dpt_pseudotime', 'leiden_orig', 'logk', 'net_prolif', 'log10SR', 'log_density_at_E3', 'log_density_at_E7', 'log_density_at_E12', 'log_density_at_E12_clip', 'log_density_at_E3_clip', 'log_density_at_E7_clip', 'log_density_at_E27', 'log_density_at_E27_clip', 'log_density_at_E49', 'log_density_at_E49_clip', 'log_density_at_E76', 'log_density_at_E76_clip', 'log_density_at_E112', 'log_density_at_E112_clip', 'log_density_at_E161', 'log_density_at_E161_clip', 'log_density_at_E269', 'log_density_at_E269_clip', 'pseudo_bulk', 'rs50', 'split', 'fine_anno'\n",
       "    var: 'symbol', 'gene_ids-10x', 'feature_types-10x', 'genome-10x', 'n_counts-10x', 'highly_variable-10x', 'means-10x', 'dispersions-10x', 'dispersions_norm-10x', 'gene_removed-10x', 'mean-10x', 'std-10x', 'geneid-SS2', 'chromosome-SS2', 'is_mito-SS2', 'highly_variable', 'mean', 'std'\n",
       "    uns: 'DM', 'DM_EigenValues', 'anno_man_colors', 'biosample_id_colors', 'data_type_colors', 'density_predictor', 'diffmap_evals', 'fine_anno_colors', 'iroot', 'isroot_colors', 'leiden', 'leiden_DM_colors', 'leiden_DM_sizes', 'leiden_colors', 'leiden_sizes', 'neighbors', 'original_pop', 'paga', 'pagaDM', 'pagaPCA', 'pca', 'phase_colors', 'pop', 'predicted_doublets_colors', 'pseudo_bulk', 'pseudo_bulk_settings', 'rank_genes_groups', 'rs50', 'umap_init_pos', 'umap_init_pos_DM'\n",
       "    obsm: 'DM_EigenVectors', 'DM_scaled', 'Delta_DM', 'TIGON_u', 'X_diffmap', 'X_pca', 'X_pca_harmony', 'X_pca_scaled', 'X_umap', 'X_umap_2d', 'X_umap_3d', 'X_umap_DM_2d', 'X_umap_DM_3d', 'mellon_u', 'surrogate_u'\n",
       "    varm: 'PCs'\n",
       "    obsp: 'DM_Kernel', 'DM_Similarity', 'DM_connectivities', 'DM_distances', 'connectivities', 'distances'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = sc.read_h5ad('data/tom_pos.h5ad')\n",
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "326eb57c",
   "metadata": {},
   "source": [
    "# 1. Gaussian KDE\n",
    "\n",
    "This is the conventional density estimation method. In the manuscript, we benchmarked several methods, and this traditional Gaussian KDE works stably and can reproduce the shift in dense cell states along the differrentaition trajectory.\n",
    "\n",
    "Gaussian KDE is used by default when runing the training `script main_train.py`. Two args can be passed to adjust\n",
    "- `--bw` : the kernel size \n",
    "- `--knn_volume`  : scale density by a cell's KNN volumn (unstable)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "146e45fc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(394, 5)\n"
     ]
    }
   ],
   "source": [
    "# list cell by time\n",
    "obs = adata.obs\n",
    "CB_ay = [obs.query('`timepoint_tx_days` == @t').index for t in adata.uns['pop']['t']]\n",
    "\n",
    "\n",
    "# cell state coordiate\n",
    "\n",
    "cellstate_key = \"DM_scaled\"\n",
    "n_dim = 5\n",
    "\n",
    "# scale to 0-1\n",
    "DM_x = adata.obsm[cellstate_key][:,:n_dim]\n",
    "\n",
    "DM_min = DM_x.min(axis=0, keepdims=True)\n",
    "DM_x_range = (DM_x.max(axis=0, keepdims=True) - DM_min)\n",
    "\n",
    "# update \n",
    "norm_DMx = (DM_x - DM_min) / DM_x_range\n",
    "adata.obsm[cellstate_key] = norm_DMx\n",
    "\n",
    "# calculate\n",
    "DM_normed = [adata[cbs].obsm[cellstate_key] for cbs in CB_ay]\n",
    "print(DM_normed[0].shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "acd36475",
   "metadata": {},
   "outputs": [],
   "source": [
    "# bandwidth\n",
    "bw = None\n",
    "bw = 0.5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4540778",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import gaussian_kde\n",
    "\n",
    "kde_kernel = []\n",
    "kde_u = []\n",
    "\n",
    "for dm in DM_normed:\n",
    "    # print(dm.shape)\n",
    "    # kde_fn = gaussian_kde(dm.T, bw_method='silverman')       # take in [n_dim, n_sample]\n",
    "    kde = gaussian_kde(dm.T, bw_method=bw)\n",
    "    kde_kernel.append(kde)\n",
    "    u_t = kde(DM_x.T)  # full DM\n",
    "    kde_u.append(u_t)\n",
    "\n",
    "kde_u = np.stack(kde_u)\n",
    "\n",
    "pdp.pl.params_in_umap(adata, kde_u, param='KDE u')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75fa48f5",
   "metadata": {},
   "source": [
    "# 2. Mellon\n",
    "\n",
    "Mellon use a gaussian process to model the cell density. See [Otto, D.J., Jordan, C., Dury, B. et al. 2024](https://www.nature.com/articles/s41592-024-02302-w) for the manuscript and [mellon github](https://github.com/settylab/mellon) for the package. \n",
    "\n",
    "\n",
    "Our package envelope the Mellon model in a simple way.\n",
    "- `pdp.tl.compute_mellon_u`\n",
    "- `pdp.tl.compute_mellon_timesense_u`\n",
    "\n",
    "***pseudodyanmics+*** integrate mellon for solving sc population dynamics. Use a different training script  `dudt_train_mellon.py`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c727bb0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For time sensitive model as an example\n",
    "\n",
    "cellstate_key = 'DM_EigenVectors'\n",
    "timepoint_key = 'timepoint_tx_days'    # \n",
    "\n",
    "log_u, mellon_fns = pdp.tl.compute_mellon_u(adata, \n",
    "                            cellstate_key = cellstate_key,\n",
    "                            timepoint_key = timepoint_key, \n",
    "                            n_dimension=None)\n",
    "\n",
    "# like KDE, each time point has its own density function \n",
    "print(\"# of estimators : \", len(mellon_fns))\n",
    "\n",
    "# a big difference is that Mellon returns log density\n",
    "pdp.pl.params_in_umap(adata, log_u, params='mellon log u')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e688ed8",
   "metadata": {},
   "source": [
    "To use time-sensitive mellon model, we need to specify the `ls_time`  parameter, which prefers short and identical interval between timepoints. For our long-term in vivo dataset, we use take the log-time for demonstration (which is a strong assumption)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4195b2d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['log_timepoint'] = adata.obs['timepoint_tx_days'].apply(lambda x: np.log(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61bf1cc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For time sensitive model as an example\n",
    "\n",
    "cellstate_key = 'DM_EigenVectors'\n",
    "timepoint_key = 'log_timepoint'    # \n",
    "\n",
    "log_u, density_predictor = pdp.tl.compute_mellon_timesense_u(adata, \n",
    "                            cellstate_key = cellstate_key,\n",
    "                            timepoint_key = timepoint_key, \n",
    "                            n_dimension=None)\n",
    "\n",
    "# this time-dependent predictor allows us to predict density at any time point\n",
    "\n",
    "t = np.log(46) # an \n",
    "X = adata.obsm[cellstate_key] \n",
    "density_d46 = density_predictor(X, t)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "def6cd0b",
   "metadata": {},
   "source": [
    "We can save this precomputed predictor for later use during training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa9250f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "density_predictor.to_json(f\"data/tom_pos_mellon_timecontinuous_predictor.json\") "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "060222cf",
   "metadata": {},
   "source": [
    "# 3. Efficient kernel density estimator for atlas level data\n",
    "\n",
    "Traditionaly gaussidan KDE scales badly to sample size. For atals with millions of cells, we can use a hash-based estimator for sample-efficiency.   \n",
    "The package `efficient_kde` provide several estimators for accelaration. see the [original github](https://github.com/talwagner/efficient_kde) for details.   \n",
    "The manuscript by [Moses Charikar and Paris Siminelakis FOCS 2017](https://arxiv.org/abs/1808.10530)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42297bd4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from efficient_kde.kde import FastLaplacianKDE, kde\n",
    "\n",
    "# Hyperparameter for HBE\n",
    "bandwidth = 1\n",
    "L = 90"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb0daa2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# list cell by time\n",
    "obs = adata.obs\n",
    "CB_ay = [obs.query('`timepoint_tx_days` == @t').index for t in adata.uns['pop']['t']]\n",
    "\n",
    "\n",
    "# cell state coordiate\n",
    "# scale to 0-1\n",
    "DM_x = adata.obsm['DM_EigenVectors_multiscaled']\n",
    "\n",
    "DM_min = DM_x.min(axis=0, keepdims=True)\n",
    "DM_x_range = (DM_x.max(axis=0, keepdims=True) - DM_min)\n",
    "\n",
    "# update \n",
    "norm_DMx = (DM_x - DM_min) / DM_x_range\n",
    "adata.obsm['DM_EigenVectors_multiscaled'] = norm_DMx\n",
    "\n",
    "# calculate\n",
    "DM_normed = [adata[cbs].obsm['DM_EigenVectors_multiscaled'] for cbs in CB_ay]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6af23b2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "hbe_kernel = []\n",
    "hbe_u = []\n",
    "\n",
    "for dm in DM_normed:\n",
    "    # print(dm.shape)\n",
    "    # kde_fn = gaussian_kde(dm.T, bw_method='silverman')       # take in [n_dim, n_sample]\n",
    "    hbe = FastLaplacianKDE(dm, bandwidth=0.05, L=30)\n",
    "    hbe_kernel.append(hbe.kde)\n",
    "    u_t = hbe.kde(DM_x)  # full DM\n",
    "    hbe_u.append(u_t)\n",
    "\n",
    "hbe_u = np.stack(hbe_u)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cec16cb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdp.pl.params_in_umap(adata, hbe_u, param='HBE u')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d15ae70f",
   "metadata": {},
   "source": [
    "# What's next \n",
    " \n",
    "- set up training configuration\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "PINN_clean",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
